File: frontend/src/types/repository.ts
// DevDocsAI — TypeScript Types: Repository
export type RepositoryStatus =
  | 'pending'
  | 'cloning'
  | 'parsing'
  | 'embedding'
  | 'graph'
  | 'summarizing'
  | 'ready'
  | 'failed';

export interface RepositoryMetadata {
  summary?: string;
  framework?: string;
  architecture_type?: string;
  total_files: number;
  total_functions: number;
  total_classes: number;
  total_lines: number;
  languages_detected?: string;
}

export interface Repository {
  id: string;
  repo_url: string;
  repo_name?: string;
  owner?: string;
  status: RepositoryStatus;
  progress: number;
  current_stage?: string;
  primary_language?: string;
  error_message?: string;
  created_at: string;
  processed_at?: string;
  metadata_?: RepositoryMetadata;
}

export interface RepositoryStatusResponse {
  repo_id: string;
  status: RepositoryStatus;
  progress: number;
  current_stage?: string;
  error_message?: string;
}

export interface AnalyzeResponse {
  repo_id: string;
  status: string;
  message: string;
}

export interface DependencyGraph {
  repo_id: string;
  mermaid: string;
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface GraphNode {
  id: string;
  name: string;
  type: string;
  file?: string;
}

export interface GraphEdge {
  source: string;
  target: string;
  relation?: string;
}

export interface ArchitectureSummary {
  repo_id: string;
  summary: string;
  components: ArchitectureComponent[];
  framework?: string;
  architecture_type?: string;
}

export interface ArchitectureComponent {
  name: string;
  type: string;
  description?: string;
  dependencies: string[];
}

File: backend/app/schemas/repository.py
"""
DevDocsAI — Pydantic Schemas (Repository)
"""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, HttpUrl, field_validator


# ── Request ──────────────────────────────────────────────────────────────────

class RepositoryAnalyzeRequest(BaseModel):
    repo_url: str

    @field_validator("repo_url")
    @classmethod
    def validate_github_url(cls, v: str) -> str:
        v = v.strip().rstrip("/")
        if not v.startswith(("https://github.com/", "http://github.com/")):
            raise ValueError("Only GitHub URLs are supported (https://github.com/...)")
        return v


# ── Response ─────────────────────────────────────────────────────────────────

class RepositoryMetadataOut(BaseModel):
    summary: Optional[str] = None
    framework: Optional[str] = None
    architecture_type: Optional[str] = None
    total_files: int = 0
    total_functions: int = 0
    total_classes: int = 0
    total_lines: int = 0
    languages_detected: Optional[str] = None

    model_config = {"from_attributes": True}


class RepositoryOut(BaseModel):
    id: str
    repo_url: str
    repo_name: Optional[str] = None
    owner: Optional[str] = None
    status: str
    progress: int
    current_stage: Optional[str] = None
    primary_language: Optional[str] = None
    error_message: Optional[str] = None
    created_at: datetime
    processed_at: Optional[datetime] = None
    metadata_: Optional[RepositoryMetadataOut] = None

    model_config = {"from_attributes": True}


class RepositoryStatusOut(BaseModel):
    repo_id: str
    status: str
    progress: int
    current_stage: Optional[str] = None
    error_message: Optional[str] = None

    model_config = {"from_attributes": True}


class RepositoryAnalyzeResponse(BaseModel):
    repo_id: str
    status: str
    message: str


class RepositoryListOut(BaseModel):
    repositories: list[RepositoryOut]
    total: int
